File: app/components/folio/console/tiptap/simple_form_wrap_component.rb
# frozen_string_literal: true

class Folio::Console::Tiptap::SimpleFormWrapComponent < Folio::Console::ApplicationComponent
  attr_reader :f

  def initialize(simple_form_model:, simple_form_options: {})
    @simple_form_model = simple_form_model
    @simple_form_options = simple_form_options

    class_names = []

    if @simple_form_options && @simple_form_options[:class].present?
      class_names += @simple_form_options[:class].split(" ")
    end

    class_names << "f-c-tiptap-simple-form-wrap__form"

    @simple_form_options[:html] ||= {}
    @simple_form_options[:html][:class] = class_names.join(" ")
  end

  def model_class
    model = @simple_form_model.is_a?(Array) ? @simple_form_model.last : @simple_form_model
    model.class
  end

  def folio_tiptap_locales
    model_class.folio_tiptap_locales
  end

  def all_tiptap_fields
    model_class.folio_tiptap_fields
  end

  def model
    @model ||= @simple_form_model.is_a?(Array) ? @simple_form_model.last : @simple_form_model
  end

  def cookie_key
    return nil unless model.respond_to?(:id) && model.id.present?

    "tiptap_attr_#{model.class.table_name}_#{model.id}"
  end

  def new_record_cookie_key
    "tiptap_attr_new_#{model.class.table_name}"
  end

  def selected_attribute
    # Check record-specific cookie first
    if cookie_key.present?
      cookie_value = controller.send(:cookies)[cookie_key.to_sym]
      if cookie_value.present? && all_tiptap_fields.include?(cookie_value)
        return cookie_value
      end
    end

    # Fallback to generic cookie for new records
    generic_cookie_value = controller.send(:cookies)[new_record_cookie_key.to_sym]
    if generic_cookie_value.present? && all_tiptap_fields.include?(generic_cookie_value)
      return generic_cookie_value
    end

    # Default to first field
    all_tiptap_fields.first
  end

  def grouped_tiptap_fields_for_locale_switcher
    locales = folio_tiptap_locales
    return [] if locales.empty?

    locales.map do |base_field, locale_array|
      {
        attribute_names: locale_array.map { |locale| "#{base_field}_#{locale}" },
        locales: locale_array
      }
    end
  end

  def data
    actions = {
      "f-input-tiptap:updateWordCount" => "updateWordCount",
      "f-c-tiptap-simple-form-wrap-autosave-info:continueUnsavedChanges" => "onContinueUnsavedChanges",
      "f-input-tiptap:tiptapContinueUnsavedChanges" => "onTiptapContinueUnsavedChanges",
      "f-input-tiptap:tiptapAutosaveFailed" => "onTiptapAutosaveFailed",
      "f-input-tiptap:tiptapAutosaveSucceeded" => "onTiptapAutosaveSucceeded",
      "f-c-file-placements-multi-picker-fields:addToPicker" => "onAddToMultiPicker",
      "f-c-file-placements-multi-picker-fields:hookOntoFormWrap" => "onMultiPickerHookOntoFormWrap",
      "f-c-ui-tabs:shown" => "onTabsChange",
      "f-c-tiptap-simple-form-wrap-locale-switch:attributeChanged" => "onAttributeChanged",
    }

    # Add form submit handler for new records
    if model.respond_to?(:new_record?) && model.new_record?
      actions["submit"] = "onFormSubmit"
    end

    stimulus_controller("f-c-tiptap-simple-form-wrap",
                        values: {
                          scrolled_to_bottom: false,
                          cookie_key: cookie_key,
                          new_record_cookie_key: new_record_cookie_key,
                        },
                        action: actions)
  end
end

File: app/components/folio/console/tiptap/simple_form_wrap_component.js
window.Folio.Stimulus.register('f-c-tiptap-simple-form-wrap', class extends window.Stimulus.Controller {
  static targets = ['scrollIco', 'scroller', 'wordCount', 'fields', 'attributeWrap']

  static values = {
    scrolledToBottom: Boolean,
    cookieKey: String,
    newRecordCookieKey: String
  }

  connect () {
    this.onScroll = window.Folio.throttle(this.onScrollRaw.bind(this))

    // Initialize current attribute name from visible attribute wrap
    const visibleWrap = this.attributeWrapTargets.find(wrap => !wrap.hidden)
    if (visibleWrap) {
      this.currentAttributeName = visibleWrap.dataset.attributeName
    }
  }

  disconnect () {
    delete this.onScroll
  }

  onScrollRaw (e) {
    const scroller = e.target
    this.scrolledToBottomValue = scroller.scrollHeight - scroller.scrollTop <= scroller.clientHeight + 1
  }

  onScrollTriggerClick (e) {
    e.preventDefault()

    const scroller = this.scrollerTarget

    if (this.scrolledToBottomValue) {
      scroller.scrollTo({
        top: 0,
        behavior: 'smooth'
      })
    } else {
      scroller.scrollTo({
        top: scroller.scrollHeight,
        behavior: 'smooth'
      })
    }
  }

  updateWordCount (e) {
    const { wordCount, attributeName } = e.detail || {}
    if (!wordCount) return
    if (!this.hasWordCountTarget) return
    this.wordCountTargets.forEach(target => {
      const wrap = target.closest('.f-c-tiptap-simple-form-wrap__attribute-wrap')
      if (wrap?.dataset.attributeName === attributeName) {
        target.dispatchEvent(new CustomEvent('f-c-tiptap-simple-form-wrap:updateWordCount', { detail: { wordCount, attributeName } }))
      }
    })
  }

  onContinueUnsavedChanges (e) {
    const tiptapInput = this.element.querySelector('[data-controller*="f-input-tiptap"]')

    if (tiptapInput) {
      tiptapInput.dispatchEvent(new CustomEvent('f-c-tiptap-simple-form-wrap:tiptapContinueUnsavedChanges'))
    }
  }

  callAutosaveInfoMethod (methodName) {
    const autosaveInfoComponent = this.element.querySelector('[data-controller*="f-c-tiptap-simple-form-wrap-autosave-info"]')

    if (autosaveInfoComponent) {
      const autosaveController = this.application.getControllerForElementAndIdentifier(autosaveInfoComponent, 'f-c-tiptap-simple-form-wrap-autosave-info')
      if (autosaveController && autosaveController[methodName]) {
        autosaveController[methodName]()
      }
    }
  }

  onTiptapContinueUnsavedChanges (e) {
    this.callAutosaveInfoMethod('hideUnsavedChanges')
  }

  onTiptapAutosaveFailed (e) {
    this.callAutosaveInfoMethod('showFailedToSave')
  }

  onTiptapAutosaveSucceeded (e) {
    this.callAutosaveInfoMethod('hideFailedToSave')
  }

  onAddToMultiPicker (e) {
    const multiPickers = this.element.querySelectorAll('.f-c-file-placements-multi-picker-fields')

    if (multiPickers.length > 1) {
      throw new Error('More than 1 multi pickers found')
    } else if (multiPickers.length === 0) {
      throw new Error('No multi pickers found')
    }

    multiPickers[0].dispatchEvent(new CustomEvent('f-c-file-placements-multi-picker-fields:addToPicker', {
      detail: e.detail
    }))
  }

  onMultiPickerHookOntoFormWrap (e) {
    const source = e.detail.source

    const wrapper = document.createElement('div')

    // add containter-fluid to extend index filters
    wrapper.className = 'f-c-tiptap-simple-form-wrap__multi-picker-wrap container-fluid'

    wrapper.appendChild(source)
    this.fieldsTarget.insertAdjacentElement('afterend', wrapper)

    this.multiPickerHooked = true
    this.checkTabsForMultiPicker()
  }

  onTabsChange () {
    this.checkTabsForMultiPicker()
  }

  checkTabsForMultiPicker () {
    if (!this.multiPickerHooked) return

    const activeLink = this.element.querySelector('.f-c-ui-tabs__nav-link.active')
    const visible = activeLink.classList.contains('f-c-file-placements-multi-picker-fields-nav-link')
    this.element.classList.toggle('f-c-tiptap-simple-form-wrap--multi-picker-visible', visible)
  }

  onAttributeChanged (e) {
    const { attributeName } = e.detail

    // Track current attribute name for new records
    this.currentAttributeName = attributeName

    // Save to cookie if cookie key is available
    if (this.cookieKeyValue) {
      const inOneDay = new Date(new Date().getTime() + 24 * 60 * 60 * 1000)
      window.Cookies.set(this.cookieKeyValue, attributeName, { expires: inOneDay, path: '' })
    }

    // Hide all attribute wraps (editors and autosave components)
    this.attributeWrapTargets.forEach(wrap => {
      wrap.hidden = wrap.dataset.attributeName !== attributeName
    })

    // Trigger resize on the visible iframe
    const visibleWrap = this.attributeWrapTargets.find(wrap => wrap.dataset.attributeName === attributeName)
    if (visibleWrap) {
      const iframe = visibleWrap.querySelector('.f-input-tiptap__iframe')
      if (iframe && iframe.contentWindow) {
        // Dispatch resize event to iframe
        iframe.contentWindow.postMessage({
          type: 'f-input-tiptap:window-resize',
          windowWidth: window.innerWidth
        }, window.origin)

        // Also trigger a resize event on the window
        setTimeout(() => {
          window.dispatchEvent(new Event('resize'))
        }, 100)
      }
    }
  }

  onFormSubmit (e) {
    // Set short-lived generic cookie for new records on form submit
    if (this.newRecordCookieKeyValue && this.currentAttributeName) {
      const inFifteenSeconds = new Date(new Date().getTime() + 15 * 1000)
      window.Cookies.set(this.newRecordCookieKeyValue, this.currentAttributeName, { expires: inFifteenSeconds, path: '' })
    }
  }
})
